backends/gpu/lib/wrapper/miopen_wrapper.cc

```diff
@@ -369,8 +370,11 @@ llvm::Error MiopenPoolingBackward(
     const miopenTensorDescriptor_t dy_desc, Pointer<const void> dy,
     const miopenTensorDescriptor_t x_desc, Pointer<const void> x,
     Pointer<const void> beta, const miopenTensorDescriptor_t dx_desc,
-    Pointer<void> dx, Pointer<void> workspace) {
+    Pointer<void> dx) {
   CheckHipContext(current);
+  size_t workspace_size;
+  RETURN_IF_ERROR(miopenPoolingGetWorkSpaceSize(y_desc, &workspace_size));
+  Pointer<void> workspace = MemAlloc(current, workspace_size).get().get();
```

Comment:
We generally want to have allocations explicitly in the IR, not hidden in a API wrapper that is supposed to not abstract things away (and cause performance problems because we can't hoist the allocation).

I also see two issues with the implementation: the `MemAlloc()` result is not checked for errors, and the result is destroyed (deallocated) before calling `miopenPoolingBackward()`.

Reply:
Sorry, I should have given more specific advice. 

I would recomment removing the `DnnPoolingBackward()` function (because cudnn and miopen have different APIs). 

Having the workspace query + allocation in the kernel implementation (`DnnPoolingBackward()` in dnn_kernels.cc) seems not too bad. Basically do a

```
if (platform == ROCm) {
  MiopenPoolingGetWorkSpaceSize()
  MemAlloc()
  MiopenPoolingBackward()
} else {
  CudnnPoolingBackward()
}
```